<!--BOOK_INFORMATION-->
<img align="left" style="padding-right:10px;" src="https://github.com/jakevdp/PythonDataScienceHandbook/blob/master/notebooks/figures/PDSH-cover-small.png?raw=1">

*This notebook contains an excerpt from the [Python Data Science Handbook](http://shop.oreilly.com/product/0636920034919.do) by Jake VanderPlas; the content is available [on GitHub](https://github.com/jakevdp/PythonDataScienceHandbook).*

*The text is released under the [CC-BY-NC-ND license](https://creativecommons.org/licenses/by-nc-nd/3.0/us/legalcode), and code is released under the [MIT license](https://opensource.org/licenses/MIT). If you find this content useful, please consider supporting the work by [buying the book](http://shop.oreilly.com/product/0636920034919.do)!*

# Aggregations: Min, Max, and Everything In Between

Often when faced with a large amount of data, a first step is to compute summary statistics for the data in question.
Perhaps the most common summary statistics are the mean and standard deviation, which allow you to summarize the "typical" values in a dataset, but other aggregates are useful as well (the sum, product, median, minimum and maximum, quantiles, etc.).

NumPy has fast built-in aggregation functions for working on arrays; we'll discuss and demonstrate some of them here.

## Summing the Values in an Array

As a quick example, consider computing the sum of all values in an array.
Python itself can do this using the built-in ``sum`` function:

In [2]:
import numpy as np

In [3]:
L = np.random.random(100)
print(L)
sum(L)

[0.72890593 0.74518574 0.19621042 0.75248906 0.01350788 0.31741442
 0.53667791 0.454839   0.38886595 0.97077062 0.72430782 0.77605674
 0.39130749 0.34292717 0.93619411 0.65513934 0.74093489 0.33008441
 0.86160953 0.85959103 0.17091335 0.46520156 0.84437169 0.10561987
 0.22583182 0.93453412 0.65946768 0.36230003 0.40123569 0.62741756
 0.78909674 0.13214606 0.70332297 0.85531618 0.74197456 0.15645294
 0.04641701 0.57735243 0.13120631 0.87969995 0.19346802 0.33637544
 0.97032347 0.71646116 0.48327299 0.15132905 0.44556015 0.25094312
 0.43493972 0.18843868 0.66189091 0.38768974 0.26740958 0.8411657
 0.41682494 0.42774774 0.98941604 0.28958155 0.76916945 0.70666485
 0.71544854 0.55345496 0.44130571 0.41308986 0.62160042 0.67730253
 0.82291537 0.15747583 0.19791879 0.02058547 0.35216482 0.82796284
 0.47291949 0.07552156 0.74281199 0.03674605 0.27098033 0.12285868
 0.66208599 0.4588969  0.75669614 0.00809009 0.67969667 0.44764188
 0.78325188 0.82457011 0.90057681 0.8789345  0.43162076 0.17210

51.0041315516985

The syntax is quite similar to that of NumPy's ``sum`` function, and the result is the same in the simplest case:

In [4]:
np.sum(L)

51.0041315516985

Be careful, though: the ``sum`` function and the ``np.sum`` function are not identical, which can sometimes lead to confusion!
In particular, their optional arguments have different meanings, and ``np.sum`` is aware of multiple array dimensions, as we will see in the following section.

## Minimum and Maximum

Similarly, Python has built-in ``min`` and ``max`` functions, used to find the minimum value and maximum value of any given array:

In [6]:
min(L), max(L)

(0.008090092149133477, 0.9894160417993104)

NumPy's corresponding functions have similar syntax [usually operate much more quickly]

In [7]:
np.min(L), np.max(L)

(0.008090092149133477, 0.9894160417993104)

For ``min``, ``max``, ``sum``, and several other NumPy aggregates, a shorter syntax is to use methods of the array object itself:

In [8]:
print(L.min(), L.max(), L.sum())

0.008090092149133477 0.9894160417993104 51.0041315516985


Whenever possible, make sure that you are using the NumPy version of these aggregates when operating on NumPy arrays!

### Multi dimensional aggregates

One common type of aggregation operation is an aggregate along a row or column.
Say you have some data stored in a two-dimensional array:

In [9]:
M = np.random.random((3, 4))
print(M)

[[0.28855397 0.78317493 0.25312424 0.93677451]
 [0.05007482 0.88107329 0.01051341 0.02723683]
 [0.07434337 0.6464296  0.18002979 0.34225116]]


By default, each NumPy aggregation function will return the aggregate over the entire array:

In [10]:
M.sum()

4.473579914531514

Aggregation functions take an additional argument specifying the *axis* along which the aggregate is computed. For example, we can find the minimum value within each column by specifying ``axis=0``:

In [11]:
M.min(axis=0)

array([0.05007482, 0.6464296 , 0.01051341, 0.02723683])

The function returns four values, corresponding to the four columns of numbers.

Similarly, we can find the maximum value within each row:

In [12]:
M.max(axis=1)

array([0.93677451, 0.88107329, 0.6464296 ])

The way the axis is specified here can be confusing to users coming from other languages.
The ``axis`` keyword specifies the *dimension of the array that will be collapsed*, rather than the dimension that will be returned.
So specifying ``axis=0`` means that the first axis will be collapsed: for two-dimensional arrays, this means that values within each column will be aggregated.

### Other aggregation functions

NumPy provides many other aggregation functions, but we won't discuss them in detail here.
Additionally, most aggregates have a ``NaN``-safe counterpart that computes the result while ignoring missing values, which are marked by the special IEEE floating-point ``NaN`` value (for a fuller discussion of missing data, see [Handling Missing Data](03.04-Missing-Values.ipynb)).
Some of these ``NaN``-safe functions were not added until NumPy 1.8, so they will not be available in older NumPy versions.

The following table provides a list of useful aggregation functions available in NumPy:

|Function Name      |   NaN-safe Version  | Description                                   |
|-------------------|---------------------|-----------------------------------------------|
| ``np.sum``        | ``np.nansum``       | Compute sum of elements                       |
| ``np.prod``       | ``np.nanprod``      | Compute product of elements                   |
| ``np.mean``       | ``np.nanmean``      | Compute mean of elements                      |
| ``np.std``        | ``np.nanstd``       | Compute standard deviation                    |
| ``np.var``        | ``np.nanvar``       | Compute variance                              |
| ``np.min``        | ``np.nanmin``       | Find minimum value                            |
| ``np.max``        | ``np.nanmax``       | Find maximum value                            |
| ``np.argmin``     | ``np.nanargmin``    | Find index of minimum value                   |
| ``np.argmax``     | ``np.nanargmax``    | Find index of maximum value                   |
| ``np.median``     | ``np.nanmedian``    | Compute median of elements                    |
| ``np.percentile`` | ``np.nanpercentile``| Compute rank-based statistics of elements     |
| ``np.any``        | N/A                 | Evaluate whether any elements are true        |
| ``np.all``        | N/A                 | Evaluate whether all elements are true        |



## Loading a file to numpy array

Aggregates available in NumPy can be extremely useful for summarizing a set of values.
As a simple example, let's consider the heights of all US presidents.
This data is available in the file *president_heights.csv*, which is a simple comma-separated list of labels and values:

In [25]:
!head -4 president_heights.txt

'head' is not recognized as an internal or external command,
operable program or batch file.


Use numpy `loadtxt` read the file and extract this information (note that the heights are measured in centimeters).
- `?np.loadtxt` can be used to explore the function `loadtxt` arguments
- use delimiter ',' since columns are separated by comma
- read first and third column
- skip the first row since it is the header

In [19]:
data = np.loadtxt('president_heights.txt',delimiter=",",skiprows=1,usecols=(0,2))
print(data.shape)
data

(42, 2)


array([[  1., 189.],
       [  2., 170.],
       [  3., 189.],
       [  4., 163.],
       [  5., 183.],
       [  6., 171.],
       [  7., 185.],
       [  8., 168.],
       [  9., 173.],
       [ 10., 183.],
       [ 11., 173.],
       [ 12., 173.],
       [ 13., 175.],
       [ 14., 178.],
       [ 15., 183.],
       [ 16., 193.],
       [ 17., 178.],
       [ 18., 173.],
       [ 19., 174.],
       [ 20., 183.],
       [ 21., 183.],
       [ 23., 168.],
       [ 25., 170.],
       [ 26., 178.],
       [ 27., 182.],
       [ 28., 180.],
       [ 29., 183.],
       [ 30., 178.],
       [ 31., 182.],
       [ 32., 188.],
       [ 33., 175.],
       [ 34., 179.],
       [ 35., 183.],
       [ 36., 193.],
       [ 37., 182.],
       [ 38., 183.],
       [ 39., 177.],
       [ 40., 185.],
       [ 41., 188.],
       [ 42., 188.],
       [ 43., 182.],
       [ 44., 185.]])

Now that we have this data array, we can compute a variety of summary statistics:

In [21]:
heights=data[:,1].copy()
print(heights)
print("Mean height:       ", heights.mean())
print("Standard deviation:", heights.std())
print("Minimum height:    ", heights.min())
print("Maximum height:    ", heights.max())

[189. 170. 189. 163. 183. 171. 185. 168. 173. 183. 173. 173. 175. 178.
 183. 193. 178. 173. 174. 183. 183. 168. 170. 178. 182. 180. 183. 178.
 182. 188. 175. 179. 183. 193. 182. 183. 177. 185. 188. 188. 182. 185.]
Mean height:        179.73809523809524
Standard deviation: 6.931843442745892
Minimum height:     163.0
Maximum height:     193.0


Note that in each case, the aggregation operation reduced the entire array to a single summarizing value, which gives us information about the distribution of values.
We may also wish to compute quantiles:

In [23]:
print("25th percentile:   ", np.percentile(heights, 25))
print("Median:            ", np.median(heights))
print("75th percentile:   ", np.percentile(heights, 75))

25th percentile:    174.25
Median:             182.0
75th percentile:    183.0


We see that the median height of US presidents is 182 cm, or just shy of six feet.
